heron/common/src/python/basics/gateway_looper.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2016 Twitter. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+'''gateway_looper.py'''
+
+import asyncore
+import errno
+import os
+import time
+import select
+
+from event_looper import EventLooper
+from heron.common.src.python.log import Log
+
+class GatewayLooper(EventLooper):
+  """A GatewayLooper, inheriting EventLooper
+
+  It is a class wrapping Python's asyncore module (and selector) to dispatch events.
+  This class can be used as a looper for an ``asyncore.dispatcher`` class, instead of calling
+  ``asyncore.loop()``.
+
+  As it is a subclass of EventLooper, it will execute in a while loop until
+  the ``exit_loop()`` is called.
+
+  In order to use this class, users first need to specify a socket map that maps from
+  a file descriptor to ``asyncore.dispatcher`` class, using ``prepare_map()`` method.
+  The GatewayLooper will dispatch ready events that are in the specified map.
+  """
+  def __init__(self):
+    super(GatewayLooper, self).__init__()
+    self.sock_map = None
+
+    # Pipe used for wake up select
+    self.pipe_r, self.pipe_w = os.pipe()
+
+    self.started = time.time()
+    Log.debug("Gateway Looper started time: " + str(time.asctime()))
+
+  def prepare_map(self, sock_map):
+    """Specifies which socket map to use; should be asyncore's socket map"""
+    self.sock_map = sock_map
```

Comment:
Could this be done in the constructor? How come a separate method to set the map?